File: src/p0017/java/solution.java
class Solution {
    public List<String> letterCombinations(String digits) {
        List<String> result = new ArrayList<>();
        if (!digits.isEmpty()) {
            dfs(digits, 0, "", result);
        }
        return result;
    }

    static final String[] n2s = new String[]{"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};

    void dfs(final String digits, int i, String s, List<String> result) {
        if (i == digits.length()) {
            result.add(s);
            return;
        }
        int n = digits.charAt(i) - '0';
        for (int p = 0; p < n2s[n].length(); p++) {
            dfs(digits, i + 1, s + n2s[n].substring(p, p + 1), result);
        }
    }
}

File: src/p0017/python/solution.py
class Solution:

    def __init__(self):
        self.n2s = ["", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"]

    def letterCombinations(self, digits: str) -> List[str]:
        result = []
        if len(digits) > 0:
            self.dfs(digits, 0, "", result)
        return result

    def dfs(self, digits, i, s, result):
        if i == len(digits):
            result.append(s)
            return
        for x in self.n2s[ord(digits[i]) - ord('0')]:
            self.dfs(digits, i + 1, s + x, result)
